presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTaskFactory.java

```diff
@@ -165,4 +183,73 @@ public RemoteTask createRemoteTask(Session session,
                 stats,
                 isBinaryTransportEnabled);
     }
+
+    @Override
+    public ListenableFuture<?> removeRemoteSource(TaskId taskId, TaskId remoteSourceTaskId)
+    {
+        URI remoteSourceUri = uriBuilderFrom(locationFactory.createLocalTaskLocation(taskId))
+                .appendPath("remote-source")
+                .appendPath(remoteSourceTaskId.toString())
+                .build();
+
+        Request request = prepareDelete()
+                .setUri(remoteSourceUri)
+                .build();
+        RequestErrorTracker errorTracker = new RequestErrorTracker(
+                taskId,
+                remoteSourceUri,
+                maxErrorDuration,
+                errorScheduledExecutor,
+                "Remove exchange remote source");
+
+        SettableFuture<?> future = SettableFuture.create();
+        removeRemoteSourceHelper(errorTracker, request, future);
+        return future;
+    }
+
+    private void removeRemoteSourceHelper(RequestErrorTracker errorTracker, Request request, SettableFuture<?> future)
+    {
+        errorTracker.startRequest();
+
+        FutureCallback<StatusResponse> callback = new FutureCallback<StatusResponse>() {
+            @Override
+            public void onSuccess(@Nullable StatusResponse response)
+            {
+                if (response == null) {
+                    throw new RuntimeException("Request failed with null response");
+                }
+                if (response.getStatusCode() != HttpStatus.OK.code()) {
+                    throw new RuntimeException("Request failed with HTTP status " + response.getStatusCode());
+                }
+                future.set(null);
+            }
+
+            @Override
+            public void onFailure(Throwable failedReason)
+            {
+                if (failedReason instanceof RejectedExecutionException && httpClient.isClosed()) {
+                    log.error("Unable to destroy exchange source at %s. HTTP client is closed", request.getUri());
```

Comment:
Let's do `future.setException` in this branch as well, to make sure that future is always resolved

Reply:
@arhimondr : hmm... I am thinking this `removeRemoteSource` should be best-effort. So if it is unable to remove, just return. And the receiver will still fail the query and we can check log to see why it failed to do failure recovery :)

Reply:
>  I am thinking this removeRemoteSource should be best-effort.

I agree with that. But still i think it is not safe to return a future that might not be resolved in some case. This will be a ticking bomb. I don't think i see the reason right now why someone would synchronize on that future (call `Future#get`), but it is in theory possible. Since `RejectedExecutionException` or `httpClient.isClosed()` are not possible, except in some very exotic conditions - it is very easy to have this bug latent for a very long period. 

So, unless there's a strong argument why shouldn't we resolve the future in this branch - i would resolve it.